ray-operator/controllers/ray/utils/ray_resources.go

```diff
@@ -0,0 +1,137 @@
+	logf "sigs.k8s.io/controller-runtime/pkg/log"
+)
+
+var rayClusterLog = logf.Log.WithName("RayCluster-Controller")
+
+// Updates the user-specified rayResource spec with CPU, GPU, and memory data from the rayStartParams and the pod template.
+// CPU, GPU, and memory data from user-specified rayResourceSpec overrides data from rayStartParams.
+// CPU, GPU, and memory data from rayStartParams overrides data from the podTemplate.
+// If CPU, GPU, or memory is not present in the resource map returned by this method, Ray will attempt to
+// infer these upon "ray start".
+// The map returned by this method is used in two ways:
+// 1. It is passed as an environment variable to the Ray node. This advertises resource capacities to the Ray scheduler.
+// 2. It is written to the RayCluster CR's Status field. This allows the Ray autoscaler to access the resource capacities.
+func ComputeRayResources(
+	rayResourceSpec rayiov1alpha1.RayResources,
+	rayStartParams map[string]string,
+	rayContainerResources v1.ResourceRequirements,
+) (updatedRayResources rayiov1alpha1.RayResources) {
+
+	updatedRayResources = make(rayiov1alpha1.RayResources)
+	// Copy user CPU, GPU, memory overrides and custom resources.
+	for key, value := range rayResourceSpec {
+		updatedRayResources[key] = value
+	}
+
+	// Compute CPU unless user has already provided an override.
+	if _, ok := updatedRayResources["CPU"]; !ok {
+		// If override is not specified, look at rayStartParams and the the Ray container's resources.
+		num_cpu := computeCPU(rayStartParams, rayContainerResources)
+		if num_cpu > 0 {
+			updatedRayResources["CPU"] = num_cpu
+		}
+	}
+
+	// Compute GPU unless user has already provided an override.
+	if _, ok := updatedRayResources["GPU"]; !ok {
+		// If override is not specified, look at rayStartParams and the the Ray container's resources.
+		num_gpu := computeGPU(rayStartParams, rayContainerResources)
+		if num_gpu > 0 {
+			updatedRayResources["GPU"] = num_gpu
+		}
+	}
+
+	// Compute memory unless user has already provided an override.
+	if _, ok := updatedRayResources["memory"]; !ok {
+		// If override is not specified, look at rayStartParams and the the Ray container's resources.
+		memory := computeMemory(rayStartParams, rayContainerResources)
+		if memory > 0 {
+			updatedRayResources["memory"] = memory
+		}
+	}
+	return updatedRayResources
+}
+
+func computeCPU(rayStartParams map[string]string, rayContainerResources v1.ResourceRequirements) int64 {
+	// Try using the num-cpus rayStartParam.
+	if cpuParam, ok := rayStartParams["num-cpus"]; ok {
+		cpuInt, err := strconv.ParseInt(cpuParam, 10, 64)
+		if err != nil {
+			rayClusterLog.Error(err, "Failed to parse num-cpus rayStartParam.")
+		} else {
+			return cpuInt
+		}
+	}
+
+	// If we couldn't read CPU from the rayStartParams, trying getting the CPU count from the
+	// container resources.
+	cpuQuantity := rayContainerResources.Limits[v1.ResourceCPU]
+	if !cpuQuantity.IsZero() {
+		return cpuQuantity.Value()
```

Comment:
Will investigate!

Reply:
@wuisawesome it rounds up to the nearest int!
Which is what we want! (to avoid the edge case of fractional CPU rounding down to 0)

Unit testing this aspect is probably not necessary (since it's documented behavior of K8s code).
Documenting is necessary!
Will start by leaving a comment in the code.

Reply:
Added comments.